src/backends/torch/torchdataaug.cc

```diff
@@ -279,11 +428,115 @@ namespace dd
+        p2.x = bbox[2]; // xmax
+        p2.y = bbox[3]; // ymax
+        origBBox.push_back(p2);
+        cv::Point2f p3;
+        p3.x = bbox[0]; // xmin
+        p3.y = bbox[3]; // ymax
+        origBBox.push_back(p3);
+        cv::Point2f p4;
+        p4.x = bbox[2]; // xmax
+        p4.y = bbox[1]; // ymin
+        origBBox.push_back(p4);
+
+        cv::perspectiveTransform(origBBox, warpedBBox, lambda);
+        float xmin = warpedBBox[0].x;
+        float ymin = warpedBBox[0].y;
+        float xmax = warpedBBox[0].x;
+        float ymax = warpedBBox[0].y;
+        for (int i = 1; i < 4; ++i)
+          {
+            if (warpedBBox[i].x < xmin)
+              xmin = warpedBBox[i].x;
+            if (warpedBBox[i].x > xmax)
+              xmax = warpedBBox[i].x;
+            if (warpedBBox[i].y < ymin)
+              ymin = warpedBBox[i].y;
+            if (warpedBBox[i].y > ymax)
+              ymax = warpedBBox[i].y;
+          }
+
+        std::vector<float> nbox = { xmin, ymin, xmax, ymax };
+        nbboxes.push_back(nbox);
+      }
+    bboxes = nbboxes;
+  }
+
+  void TorchImgRandAugCV::filterBBoxes(std::vector<std::vector<float>> &bboxes,
+                                       const int &img_width,
+                                       const int &img_height)
+  {
+    std::vector<std::vector<float>> nbboxes;
+    for (size_t i = 0; i < bboxes.size(); ++i)
+      {
+        std::vector<float> bbox = bboxes.at(i);
+        if (bbox[2] >= 0.0 && bbox[0] <= img_width && bbox[3] >= 0.0
+            && bbox[1] <= img_height)
+          {
+            std::vector<float> nbbox;
+            nbbox.push_back(std::max(0.0f, bbox[0])); // xmin
+            nbbox.push_back(std::max(0.0f, bbox[1])); // ymin
+            nbbox.push_back(
+                std::min(static_cast<float>(img_width), bbox[2])); // xmax
+            nbbox.push_back(
+                std::min(static_cast<float>(img_height), bbox[3])); // ymax
+            float surfbb = (bbox[2] - bbox[0]) * (bbox[3] - bbox[1]);
+            float surfnbb = (nbbox[2] - nbbox[0]) * (nbbox[3] - nbbox[1]);
+            if (surfnbb > 0.75 * surfbb) // keep bboxes that are at least 75%
+                                         // as big as the original
+              {
+                nbboxes.push_back(nbbox);
+              }
+          }
+      }
+    bboxes = nbboxes;
+  }
+
+  void TorchImgRandAugCV::applyGeometryBBox(
+      std::vector<std::vector<float>> &bboxes, const GeometryParams &cp,
+      const int &img_width, const int &img_height)
+  {
+    // XXX: fix (enlarged bboxes for constant padding)
+    for (size_t i = 0; i < bboxes.size(); ++i)
+      {
+        bboxes[i][0] += img_width;
+        bboxes[i][2] += img_width;
+        bboxes[i][1] += img_height;
+        bboxes[i][3] += img_height;
+      }
+
+    // use cp lambda on bboxes
+    warpBBoxes(bboxes, cp._lambda);
+
+    // filter bboxes
+    filterBBoxes(bboxes, img_width, img_height);
+    // TODO: reject if 0 bbox
```

Comment:
what happens at the moment? models are fed images with 0 bbox during training?

Reply:
My understanding is that the sample should be discarded as object detection models cannot be trained from 0 bbox.

Reply:
Two choices if the transformation yields 0 bboxes:
- reject the image
- reject the transformation

I'll implement the second choice by cancelling the effect of the geometric augmentation on that particular image.